src/url.js

```diff
@@ -146,6 +159,126 @@ function finalizeResourceType(resourceType = "image", type = "upload", urlSuffix
   return [resourceType, type].join("/");
 }
 
+/**
+ * Encode publicId
+ * @param publicId
+ * @returns {string} encoded publicId
+ */
+function encodePublicId(publicId) {
+  return encodeURIComponent(publicId).replace(/%3A/g, ':').replace(/%2F/g, '/');
+}
+
+/**
+ * Encode and format publicId
+ * @param publicId
+ * @param options
+ * @returns {string} publicId
+ */
+function formatPublicId(publicId, options) {
+  if (publicId.match(/^https?:/)) {
+    publicId = encodePublicId(publicId);
+  } else {
+    try {
+      // Make sure publicId is URI encoded.
+      publicId = decodeURIComponent(publicId);
+    } catch (error) {}
+
+    publicId = encodePublicId(publicId);
+
+    if (options.url_suffix) {
+      publicId = publicId + '/' + options.url_suffix;
+    }
+    if (options.format) {
+      if (!options.trust_public_id) {
+        publicId = publicId.replace(/\.(jpg|png|gif|webp)$/, '');
+      }
+      publicId = publicId + '.' + options.format;
+    }
+  }
+  return publicId;
+}
+
+/**
+ * Get any error with url options
+ * @param options
+ * @returns {string} if error, otherwise return undefined
+ */
+function validate(options) {
+  const {cloud_name, url_suffix} = options;
+
+  if (!cloud_name) {
+    return 'Unknown cloud_name';
+  }
+
+  if (url_suffix && url_suffix.match(/[\.\/]/)) {
+    return 'url_suffix should not include . or /';
+  }
+}
+
+/**
+ * Check if the publicId is the actual url
+ * @param publicId
+ * @param type
+ * @returns {boolean}
+ */
+function isPublicIdUrl(publicId, type) {
+  return !!(publicId.match(/^https?:/) && (type === 'upload' || type === 'asset'));
+}
+
+/**
+ * Get version part of the url
+ * @param publicId
+ * @param options
+ * @returns {string}
+ */
+function handleVersion(publicId, options) {
+  const isForceVersion = (options.force_version || typeof options.force_version === 'undefined');
+  const isPublicIdIncludesVersion = publicId.search('/') < 0 || publicId.match(/^v[0-9]+/) || publicId.match(/^https?:\//);
+  const isVersionSet = options.version && options.version.toString();
```

Comment:
@maoznir , please add some explanation about the logics here.. what does each section check?

Reply:
done

Reply:
@maoznir , thanks! much clearer